### PR TITLE
Adding CoflatMap and tests to WriterT and fixed laws for CoflatMap

### PR DESCRIPTION
**Pull request by zainab-ali**
Adding a `CoflatMap` instance to `WriterT`, as required in #620 
Added all laws in `CoflatMapLaws` to `CoflatMapTests`

### CODE REVIEW

File: tests/src/test/scala/cats/tests/WriterTTests.scala [hunk 1]
```diff
@@ -268,4 +268,15 @@ class WriterTTests extends CatsSuite {
 
     Semigroup[WriterT[Id, Int, Int]]
   }
+
+  {
+    // F has a functor
+    implicit val F: Functor[ListWrapper] = ListWrapper.functor
+
+    CoflatMap[WriterT[ListWrapper, Int, ?]]
+    checkAll("WriterT[Listwrapper, Int, ?]", CoflatMapTests[WriterT[ListWrapper, Int, ?]].coflatMap[Int, Int, Int])
```
Comment (adelbertc) [inline tests/src/test/scala/cats/tests/WriterTTests.scala]: need to check serializability

File: core/src/main/scala/cats/data/WriterT.scala [hunk 2]
```diff
@@ -87,8 +87,7 @@ private[data] sealed abstract class WriterTInstances0 extends WriterTInstances1
       implicit val L0: Monoid[L] = L
     }
 
-  implicit def writerTIdFunctor[L]: Functor[WriterT[Id, L, ?]] =
-    writerTFunctor[Id, L]
+  def writerTIdFunctor[L]: Functor[WriterT[Id, L, ?]] = writerTIdCoflatMap
```
Comment (ceedubs) [inline core/src/main/scala/cats/data/WriterT.scala]: We should be able to delete this now that `writerTIdCoflatMap` exists. You might want to move it up to where this was.

File: core/src/main/scala/cats/data/WriterT.scala [hunk 3]
```diff
@@ -168,7 +172,10 @@ private[data] sealed abstract class WriterTInstances6 extends WriterTInstances7
 }
 
 private[data] sealed abstract class WriterTInstances7 {
-  implicit def writerTFunctor[F[_], L](implicit F: Functor[F]): Functor[WriterT[F, L, ?]] = new WriterTFunctor[F, L] {
+
+  def writerTFunctor[F[_], L](implicit F: Functor[F]): Functor[WriterT[F, L, ?]] = writerTCoflatMap
```
Comment (ceedubs) [inline core/src/main/scala/cats/data/WriterT.scala]: I don't think this is needed for anything.

**Comment by ceedubs** [comment]
Thanks @zainab-ali! This is definitely a nice change. I think it allows us to remove a couple `Functor` instances and have left some inline comments in the relevant places.

File: core/src/main/scala/cats/data/WriterT.scala (hunk 2 above)
Comment (zainab-ali) [inline core/src/main/scala/cats/data/WriterT.scala]: If I move it up, the implicit conflicts with `writerTIdFlatMap` when there is a `Semigroup` for `L`, as they are both `Functors`

Comment (ceedubs) [inline core/src/main/scala/cats/data/WriterT.scala]: Ah I see. Glad we have tests for that! :). Do things work if you just remove this `writerTIdFunctor`?

File: core/src/main/scala/cats/data/WriterT.scala [hunk 4]
```diff
@@ -168,7 +171,9 @@ private[data] sealed abstract class WriterTInstances6 extends WriterTInstances7
 }
 
 private[data] sealed abstract class WriterTInstances7 {
-  implicit def writerTFunctor[F[_], L](implicit F: Functor[F]): Functor[WriterT[F, L, ?]] = new WriterTFunctor[F, L] {
+
+  implicit def writerTCoflatMap[F[_], L](implicit F: Functor[F]): WriterTCoflatMap[F, L] =
```
Comment (ceedubs) [inline core/src/main/scala/cats/data/WriterT.scala]: I think the return type here should be `CoflatMap[WriterT[F, L, ?]]`

File: tests/src/test/scala/cats/tests/WriterTTests.scala [hunk 5]
```diff
@@ -268,4 +268,16 @@ class WriterTTests extends CatsSuite {
...
+  {
+    // F has a Functor
+    implicit val F: Functor[ListWrapper] = ListWrapper.functor
+
+    CoflatMap[WriterT[ListWrapper, Int, ?]]
+    checkAll("WriterT[Listwrapper, Int, ?]", CoflatMapTests[WriterT[ListWrapper, Int, ?]].coflatMap[Int, Int, Int])
+    checkAll("WriterT[ListWrapper, Int, ?]", SerializableTests.serializable(CoflatMap[WriterT[ListWrapper, Int, ?]]))
+
+    // Id has a Functor
+    CoflatMap[WriterT[Id, Int, ?]]
```
Comment (ceedubs) [inline tests/src/test/scala/cats/tests/WriterTTests.scala]: Could you also summon the `Functor[WriterT[Id, Int, ?]]` and ``Functor[WriterT[ListWrapper, Int, ?]]` to ensure there aren't ambiguity issues with the `Functor` instances?

**Comment by ceedubs** [comment]
@zainab-ali looks great! Thanks again. 👍

**Comment by adelbertc** [comment]
LGTM 👍

**Pull request merged**